elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,105 @@
+
+The general intent for this enhancement is to develop some form of configurable expression resolver.
+
+The expression resolver will reference a specific alias within a specific credential store, this credential will be used to decrypt the expression which was previously encrypted using this credential.  The expression resolver will also contain configuration such as the algorithm to use. 
+
+The expression resolver should support both generated keys and password based encryption.
+
+Using the credential store it will be possible to either add or generate suitable keys to be used for the encryption / decryption of values.  This will need to include: -
+
+ * The standalone elytron tool.
+ * The management operations exposed on the credential store.
+
+Where generated keys are used the tools and management operations should support exporting and importing the keys to allow them to be 'moved' between credential stores.  Where possible this should be achieved using specification defined formats.
+
+Functionality should be added to allow an administrator to encrypt a clear text String so that it can be stored as an expression, this will be added to: -
+
+ * The standalone elytron tool.
+ * New management operations, this may however be against the expression resolver not the credential store. 
+
+The domain mode relationships will need to considered further, where we used the PicketBox vault we defined the vault within the host.xml and used it's values in the domain.xml.  As the credential store can both be defined in the host.xml and the domain.xml it would make sense to make use of the one stored within that descriptor, however this could be impacted based on how the configurable expression resolver is defined.
+
+=== Nice-to-Have Requirements
+
+Support for multiple expression resolvers being defined, allowing the correct one to be identified for a specific expression value.  It may not be possible to implement this in the initial enhancement however this should still be considered to allow it's potential addition at a later point.
+
+It would be nice to cross reference subsystem managed security providers for the expression resolver, however this component needs to be usable at the start of `Stage.RUNTIME` so there will be a limit as to how many subsystem managed resources can be depended upon.
+
+=== Non-Requirements
+
+This enhancement will not support the retrieval of plain text strings from the credential store, this enhancement is specifically adding support for decrypting reversibly encrypted attribute expression values.
+
+Automatic encryption of attribute values will not be supported via this enhancement, as multiple steps are required that would be better performed within enhancements to the management tooling - each of which would require special consideration based on their own user interfaces.
+
+As with other CredentialStore use cases no automatic replication of the store or it's entries are supported with this enhancement.
```

Comment:
Can you please explain to me what replication means in this context?

Reply:
In domain mode we will not automatically transfer the contents of one credential store to another node in the domain.